### PR TITLE
[PackageDescription] Add Enforced Semantic Versioning

### PR DESCRIPTION
**Pull request by JPMartha**
This is part of proposal for **[Enforced Semantic Versioning](https://github.com/apple/swift-package-manager/blob/master/Documentation/PackageManagerCommunityProposal.md#enforced-semantic-versioning)**.

First of all, I implemented `newVersionEnforcedSemanticVersioning` function and its tests, etc.
I intend to change small, so I have NOT implemented how to enforce Semantic Versioning yet.

For now, this function is called only at its tests.

===

By the way, I have some concerns about this.

1. How to handle overflow (e.g. the next to `Version(Int.max,Int.max,Int.max)`).

  I don't think it's an individual issue, so I have not implemented yet.

2. It might be implemented in other file. (e.g. `Build/PackageVersion.swift`)
3. These name are so long that it might be better to renamed them.

  e.g.
  - `newVersionEnforcedSemanticVersioning` to `newVersionEnforcedSemVer` or `nextVersionEnforcedSemVer`
  - `testNewVersionEnforcedSemanticVersioning` to `testNewVersionEnforcedSemVer` or `testNextVersionEnforcedSemVer`

### CODE REVIEW

**Comment by ddunbar** [comment]
One question here, which I'm not sure we have already addressed (and this isn't really the right forum, but it is related).

Dart went the route of extending semver semantics to pre-1.0.0 release, but with the semantics shifted. So 0.2 is major change on 0.1. See https://www.dartlang.org/tools/pub/versioning.html

I rather like this, it fits with the 0.x.x conventions well, and I like the ability to start at 0.x.x.

What do other people think? @mxcl @rballard @abertelrud?

File: Sources/PackageDescription/Version.swift [hunk 1]
```diff
@@ -116,3 +116,31 @@ extension Version: CustomStringConvertible {
...
+// MARK: Enforced Semantic Versioning
+
+enum UpdateType {
+    /// e.g. Remove a method from the public API or change a public method's signature
+    case Major
+    /// e.g. A new method is added to the public API
+    case Minor
+    /// e.g. Change only the implementation of a method
+    case Patch
+}
```
Comment (ddunbar) [inline Sources/PackageDescription/Version.swift]: The one thing I would consider changing here is the naming of the types to maybe use different names not synonymous with the traditional version component names.

If we do decide we like the Dart style 0.1.0 -> 0.2.0 is a "major" change, then it is probably a good idea to use a different name than "major" for the enum.

**Comment by JPMartha** [comment]
Thanks for your reply.
And, I'm really sorry for my late reply.

I just followed an example mentioned in [Swift Package Manager Community Proposal](https://github.com/apple/swift-package-manager/blob/master/Documentation/PackageManagerCommunityProposal.md#enforced-semantic-versioning).
I will follow the rule which we discuss with.

I'm sorry if that is not an answer.

**Comment by aciidb0mb3r** [comment]
Tracking efforts for this here https://bugs.swift.org/browse/SR-2917

**Pull request closed without merging**